File: realty/backend/app/services/score_utils.py
from __future__ import annotations

import math
import re
from dataclasses import dataclass
from datetime import date
from typing import Any


def clamp(x: float, lo: float, hi: float) -> float:
    if x is None or math.isnan(x):
        return lo
    return max(lo, min(hi, x))


_FLOOR_RE = re.compile(r"(-?\d+)")


def parse_floor_number(floor_number: str | None) -> int | None:
    """
    Parse floor number from strings like:
      - '第6楼'
      - '6楼'
      - '6'
    Returns first integer, or None if not parseable.
    """
    if not floor_number:
        return None
    m = _FLOOR_RE.search(floor_number)
    if not m:
        return None
    try:
        return int(m.group(1))
    except Exception:
        return None


def maybe_scale_trend_delta(trend_delta_raw: float) -> float:
    """
    trend_delta_raw may be:
    - fraction change in [-1, 1] (e.g. +0.12 meaning +12%)
    - or already a scaled value (e.g. +10, -3)

    We normalize to a roughly [-100, 100] range for V1 mapping.
    """
    if abs(trend_delta_raw) <= 1.0:
        return trend_delta_raw * 100.0
    return trend_delta_raw


def current_year_floor(d: date | None = None) -> int:
    d = d or date.today()
    return d.year


@dataclass
class MissingFallback:
    field: str
    policy: str
    reason: str | None = None


def json_list_max(values: list[float] | None) -> float | None:
    if not values:
        return None
    return max(values)

File: realty_app/src/rules/scoreUtils.ts
/**
 * 通用工具，对应 Python 端 `realty/backend/app/services/score_utils.py`。
 * 保持字段名与 Python 版完全一致，方便对照单测。
 */

export function clamp(x: number | null | undefined, lo: number, hi: number): number {
  if (x == null || Number.isNaN(x)) return lo;
  return Math.max(lo, Math.min(hi, x));
}

const FLOOR_RE = /(-?\d+)/;

export function parseFloorNumber(floorNumber: string | null | undefined): number | null {
  if (!floorNumber) return null;
  const m = FLOOR_RE.exec(floorNumber);
  if (!m) return null;
  try {
    return parseInt(m[1], 10);
  } catch {
    return null;
  }
}

/**
 * Python 版的 maybe_scale_trend_delta：
 * - abs <= 1.0 当作百分比，乘 100
 * - 否则当作已 scale 的值，原样返回
 */
export function maybeScaleTrendDelta(trendDeltaRaw: number): number {
  if (Math.abs(trendDeltaRaw) <= 1.0) return trendDeltaRaw * 100.0;
  return trendDeltaRaw;
}

export function currentYearFloor(d: Date | null = null): number {
  return (d ?? new Date()).getFullYear();
}

export interface MissingFallback {
  field: string;
  policy: string;
  reason: string;
}

export function jsonListMax(values: number[] | null | undefined): number | null {
  if (!values || values.length === 0) return null;
  return Math.max(...values);
}